File: .codex/hooks/latest_dependency_instruction.py
#!/usr/bin/env python3
"""Reject dependency additions that rely on an explicitly supplied version."""

import json
import os
import re
import sys

from hook_response import deny
from shell_command import SHELLS, shell_payload, shell_segments, strip_leading_shell_prefix


PYTHON_EXECUTABLE = re.compile(r"^(?:python|python\d+(?:\.\d+)?)$")
PYTHON_CONSTRAINT = re.compile(r"(?:===|==|~=|!=|<=|>=|<|>)")

INSTRUCTION = (
    "Dependency addition blocked because the command supplied an explicit version. "
    "Do not use a package version recalled from model memory. Query the authoritative "
    "package registry or package-manager metadata now, select the latest stable release "
    "compatible with explicit project and runtime constraints, and retry with an "
    "unversioned package specifier so the package manager resolves the current release. "
    "If the user or project explicitly requires an older version, explain that constraint "
    "instead of bypassing this guard."
)


def _positionals(arguments: list[str]) -> list[str]:
    return [argument for argument in arguments if not argument.startswith("-")]


def _javascript_spec_has_version(spec: str) -> bool:
    if spec.startswith((".", "/", "file:", "git:", "git+", "http:", "https:")):
        return False
    separator = spec.rfind("@")
    if separator <= 0:
        return False
    version = spec[separator + 1 :].lower()
    return version not in {"", "*", "latest", "workspace:*", "workspace:^", "workspace:~"}


def _python_spec_has_version(spec: str) -> bool:
    if spec.startswith((".", "/", "file:", "git+", "http:", "https:")):
        return False
    return PYTHON_CONSTRAINT.search(spec) is not None


def _explicit_version(executable: str, arguments: list[str]) -> bool:
    if executable in {"npm", "pnpm", "bun"}:
        if not arguments or arguments[0] not in {"add", "i", "install"}:
            return False
        return any(
            _javascript_spec_has_version(spec)
            for spec in _positionals(arguments[1:])
        )

    if executable == "yarn":
        return bool(arguments and arguments[0] == "add") and any(
            _javascript_spec_has_version(spec)
            for spec in _positionals(arguments[1:])
        )

    if executable == "deno":
        return bool(arguments and arguments[0] == "add") and any(
            _javascript_spec_has_version(spec)
            for spec in _positionals(arguments[1:])
        )

    if executable in {"pip", "pip3"}:
        return bool(arguments and arguments[0] == "install") and any(
            _python_spec_has_version(spec)
            for spec in _positionals(arguments[1:])
        )

    if PYTHON_EXECUTABLE.fullmatch(executable):
        return (
            len(arguments) >= 2
            and arguments[:2] == ["-m", "pip"]
            and _explicit_version("pip", arguments[2:])
        )

    if executable == "uv":
        if arguments[:2] == ["pip", "install"]:
            specs = arguments[2:]
        elif arguments and arguments[0] == "add":
            specs = arguments[1:]
        else:
            return False
        return any(_python_spec_has_version(spec) for spec in _positionals(specs))

    if executable == "poetry":
        return bool(arguments and arguments[0] == "add") and any(
            _javascript_spec_has_version(spec) or _python_spec_has_version(spec)
            for spec in _positionals(arguments[1:])
        )

    if executable == "pipenv":
        return bool(arguments and arguments[0] == "install") and any(
            _python_spec_has_version(spec)
            for spec in _positionals(arguments[1:])
        )

    if executable == "cargo":
        return bool(arguments and arguments[0] == "add") and (
            any(
                argument in {"--vers", "--version"}
                or argument.startswith(("--vers=", "--version="))
                for argument in arguments[1:]
            )
            or any(
                _javascript_spec_has_version(spec)
                for spec in _positionals(arguments[1:])
            )
        )

    if executable == "go":
        return bool(arguments and arguments[0] in {"get", "install"}) and any(
            _javascript_spec_has_version(spec)
            for spec in _positionals(arguments[1:])
        )

    if executable == "composer":
        return bool(arguments and arguments[0] == "require") and any(
            _javascript_spec_has_version(spec)
            or bool(re.search(r"(?::|\^|~|[<>=])\d", spec))
            for spec in _positionals(arguments[1:])
        )

    if executable in {"bundle", "gem"}:
        subcommand = {"bundle": "add", "gem": "install"}[executable]
        return bool(arguments and arguments[0] == subcommand) and (
            any(
                argument in {"-v", "--version"} or argument.startswith("--version=")
                for argument in arguments[1:]
            )
            or any(
                _javascript_spec_has_version(spec)
                or bool(re.search(r"(?:\^|~|[<>=])\d", spec))
                for spec in _positionals(arguments[1:])
            )
        )

    if executable == "dotnet":
        return arguments[:2] == ["add", "package"] and any(
            argument == "--version" or argument.startswith("--version=")
            for argument in arguments[2:]
        )

    if executable == "luarocks" and arguments and arguments[0] == "install":
        return len(_positionals(arguments[1:])) > 1

    return False


def _segment_has_explicit_dependency_version(tokens: list[str], depth: int) -> bool:
    remaining = strip_leading_shell_prefix(tokens)
    if not remaining:
        return False

    executable = os.path.basename(remaining[0])
    arguments = remaining[1:]
    if executable == "rtk":
        return _segment_has_explicit_dependency_version(arguments, depth)
    if executable in SHELLS:
        payload = shell_payload(arguments)
        return (
            has_explicit_dependency_version(payload, depth + 1)
            if payload is not None
            else False
        )
    return _explicit_version(executable, arguments)


def has_explicit_dependency_version(command: object, depth: int = 0) -> bool:
    """Return whether a shell command adds a dependency with a supplied version."""
    if not isinstance(command, str) or not command.strip() or depth > 4:
        return False
    return any(
        _segment_has_explicit_dependency_version(segment, depth)
        for segment in shell_segments(command)
    )


def response_for(event: object) -> dict[str, object] | None:
    if not isinstance(event, dict):
        return None
    tool_input = event.get("tool_input")
    if not isinstance(tool_input, dict):
        return None
    if not has_explicit_dependency_version(tool_input.get("command")):
        return None
    return deny(INSTRUCTION)


def main() -> int:
    try:
        event = json.load(sys.stdin)
    except (json.JSONDecodeError, UnicodeDecodeError):
        return 0
    response = response_for(event)
    if response is not None:
        print(json.dumps(response))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .codex/hooks/latest_dependency_instruction.ts
import * as path from "path";
import { deny } from "./lib/hook_response.ts";
import { handleVersion, isRecord, readEvent, runMain, writeJson } from "./lib/hook_runtime.ts";
import {
  SHELLS,
  shellPayload,
  shellSegments,
  stripLeadingShellPrefix,
} from "./lib/shell_command.ts";

export const VERSION = "1.0.1";

const PYTHON_EXECUTABLE = /^(?:python|python\d+(?:\.\d+)?)$/;
const PYTHON_CONSTRAINT = /(?:===|==|~=|!=|<=|>=|<|>)/;
export const INSTRUCTION =
  "Dependency addition blocked because the command supplied an explicit version. " +
  "Do not use a package version recalled from model memory. Query the authoritative " +
  "package registry or package-manager metadata now, select the latest stable release " +
  "compatible with explicit project and runtime constraints, and retry with an " +
  "unversioned package specifier so the package manager resolves the current release. " +
  "If the user or project explicitly requires an older version, explain that constraint " +
  "instead of bypassing this guard.";

function positionals(arguments_: string[]): string[] {
  return arguments_.filter((argument) => !argument.startsWith("-"));
}

function javascriptSpecHasVersion(spec: string): boolean {
  if (
    spec.startsWith(".") ||
    spec.startsWith("/") ||
    spec.startsWith("file:") ||
    spec.startsWith("git:") ||
    spec.startsWith("git+") ||
    spec.startsWith("http:") ||
    spec.startsWith("https:")
  ) {
    return false;
  }
  const separator = spec.lastIndexOf("@");
  if (separator <= 0) {
    return false;
  }
  const version = spec.slice(separator + 1).toLowerCase();
  return !["", "*", "latest", "workspace:*", "workspace:^", "workspace:~"].includes(version);
}

function pythonSpecHasVersion(spec: string): boolean {
  if (
    spec.startsWith(".") ||
    spec.startsWith("/") ||
    spec.startsWith("file:") ||
    spec.startsWith("git+") ||
    spec.startsWith("http:") ||
    spec.startsWith("https:")
  ) {
    return false;
  }
  return PYTHON_CONSTRAINT.test(spec);
}

function explicitVersion(executable: string, arguments_: string[]): boolean {
  if (["npm", "pnpm", "bun"].includes(executable)) {
    if (arguments_.length === 0 || !["add", "i", "install"].includes(arguments_[0] ?? "")) {
      return false;
    }
    return positionals(arguments_.slice(1)).some(javascriptSpecHasVersion);
  }
  if (executable === "yarn") {
    return (arguments_[0] === "add") && positionals(arguments_.slice(1)).some(javascriptSpecHasVersion);
  }
  if (executable === "deno") {
    return (arguments_[0] === "add") && positionals(arguments_.slice(1)).some(javascriptSpecHasVersion);
  }
  if (executable === "pip" || executable === "pip3") {
    return (arguments_[0] === "install") && positionals(arguments_.slice(1)).some(pythonSpecHasVersion);
  }
  if (PYTHON_EXECUTABLE.test(executable)) {
    return (
      arguments_.length >= 2 &&
      arguments_[0] === "-m" &&
      arguments_[1] === "pip" &&
      explicitVersion("pip", arguments_.slice(2))
    );
  }
  if (executable === "uv") {
    let specs: string[];
    if (arguments_[0] === "pip" && arguments_[1] === "install") {
      specs = arguments_.slice(2);
    } else if (arguments_[0] === "add") {
      specs = arguments_.slice(1);
    } else {
      return false;
    }
    return positionals(specs).some(pythonSpecHasVersion);
  }
  if (executable === "poetry") {
    return (
      arguments_[0] === "add" &&
      positionals(arguments_.slice(1)).some(
        (spec) => javascriptSpecHasVersion(spec) || pythonSpecHasVersion(spec),
      )
    );
  }
  if (executable === "pipenv") {
    return (arguments_[0] === "install") && positionals(arguments_.slice(1)).some(pythonSpecHasVersion);
  }
  if (executable === "cargo") {
    return (
      arguments_[0] === "add" &&
      (arguments_.slice(1).some(
        (argument) =>
          argument === "--vers" ||
          argument === "--version" ||
          argument.startsWith("--vers=") ||
          argument.startsWith("--version="),
      ) ||
        positionals(arguments_.slice(1)).some(javascriptSpecHasVersion))
    );
  }
  if (executable === "go") {
    return (
      (arguments_[0] === "get" || arguments_[0] === "install") &&
      positionals(arguments_.slice(1)).some(javascriptSpecHasVersion)
    );
  }
  if (executable === "composer") {
    return (
      arguments_[0] === "require" &&
      positionals(arguments_.slice(1)).some(
        (spec) => javascriptSpecHasVersion(spec) || /(?::|\^|~|[<>=])\d/.test(spec),
      )
    );
  }
  if (executable === "bundle" || executable === "gem") {
    const subcommand = executable === "bundle" ? "add" : "install";
    return (
      arguments_[0] === subcommand &&
      (arguments_.slice(1).some(
        (argument) => argument === "-v" || argument === "--version" || argument.startsWith("--version="),
      ) ||
        positionals(arguments_.slice(1)).some(
          (spec) => javascriptSpecHasVersion(spec) || /(?:\^|~|[<>=])\d/.test(spec),
        ))
    );
  }
  if (executable === "dotnet") {
    return (
      arguments_[0] === "add" &&
      arguments_[1] === "package" &&
      arguments_.slice(2).some((argument) => argument === "--version" || argument.startsWith("--version="))
    );
  }
  if (executable === "luarocks" && arguments_[0] === "install") {
    return positionals(arguments_.slice(1)).length > 1;
  }
  return false;
}

function segmentHasExplicitDependencyVersion(tokens: string[], depth: number): boolean {
  const remaining = stripLeadingShellPrefix(tokens);
  if (remaining.length === 0) {
    return false;
  }
  const executable = path.basename(remaining[0] ?? "");
  const arguments_ = remaining.slice(1);
  if (executable === "rtk") {
    return segmentHasExplicitDependencyVersion(arguments_, depth);
  }
  if (SHELLS.has(executable)) {
    const payload = shellPayload(arguments_);
    return payload !== undefined && hasExplicitDependencyVersion(payload, depth + 1);
  }
  return explicitVersion(executable, arguments_);
}

export function hasExplicitDependencyVersion(command: unknown, depth = 0): boolean {
  if (typeof command !== "string" || command.trim().length === 0 || depth > 4) {
    return false;
  }
  return shellSegments(command).some((segment) =>
    segmentHasExplicitDependencyVersion(segment, depth),
  );
}

export function responseFor(event: unknown): Record<string, unknown> | undefined {
  if (!isRecord(event)) {
    return undefined;
  }
  const toolInput = event.tool_input;
  if (!isRecord(toolInput)) {
    return undefined;
  }
  if (!hasExplicitDependencyVersion(toolInput.command)) {
    return undefined;
  }
  return deny(INSTRUCTION);
}

function main(): number {
  if (handleVersion(VERSION)) {
    return 0;
  }
  const response = responseFor(readEvent());
  if (response !== undefined) {
    writeJson(response);
  }
  return 0;
}

runMain("latest_dependency_instruction", main);
